File: Code/trunk/cpp/Misc/Arguments.cpp
/*Arguments.cpp
 $Author: jlconlin $
 $Id: Arguments.cpp 683 2009-07-07 19:37:16Z jlconlin $
 $Revision: 683 $
 $Date: 2009-07-07 13:37:16 -0600 (Tue, 07 Jul 2009) $

 */

#include <iostream>
#include <list>
#include <limits>
#include <sstream>
#include "tclap/CmdLine.h"
#include "Arguments.hpp"

using std::cout;
using std::endl;
using std::string;

Arguments::Arguments(string cmdLine)
:   cmd(cmdLine),
    tracks("t", "tracks", "# Histories to track", false, 50000, "N"),
    inactive("i", "inactive", "# Inactive Cycles", false, 0, "i"),
    active("a", "active", "# Active Cycles", false, 20, "a"),
    bins("b", "bins", "# spatial bins for discretization", false, 50, "b"),
    width("w", "width", "Width of slab", false, 20, "Width"),
    seed("s", "seed", "", false, 0, "Random seed"),
    verbose("v", "verbose", "The verbosity of the output", false, 2, "v"),
    tolerance( "", "tol", "Tolerance", false, 
            std::numeric_limits<double>::epsilon(), "tol"),
    terminal("", "terminal", "Which gnuplot terminal", false, "dumb", "term"),
    MultiMedia("m", "multimedia", "Multi Media problem", false),
    filename("f", "filename", "Filename for data", true, "data.dat", "f"),
    run("", "run", "Run simulation", false),
    _cmdLine()
{   
    // Add arguments to command line
    cmd.add(tracks);
    cmd.add(inactive);
    cmd.add(active);
    cmd.add(bins);
    cmd.add(width);
    cmd.add(seed);
    cmd.add(verbose);
    cmd.add(tolerance);
    cmd.add(terminal);
    cmd.add(MultiMedia);
    cmd.add(filename);
    cmd.add(run);
}

string Arguments::Args() {
    std::ostringstream arguments;
    arguments << "Arguments: ";

    std::list<TCLAP::Arg*> args = cmd.getArgList();
    TCLAP::ArgListIterator ArgIter = args.begin();
    for( ; ArgIter != args.end(); ++ArgIter ){
        arguments << (*ArgIter)->getName() << " = " 
            << (*ArgIter)->stringRep() << ", ";
    }

    return arguments.str();
}

// Parse command line arguments
void Arguments::Parse(int argc, char* argv[] ) { 
    std::ostringstream line;
    for( int i = 0; i < argc; ++i ) line << argv[i] << " ";
     _cmdLine = line.str();

    // Parse command line
    cmd.parse(argc, argv);
}

File: Code/trunk/cpp/Misc/Arguments.py
__id__       = "$Id: Arguments.py 684 2009-07-07 20:05:36Z jlconlin $"
__author__   = "$Author: jlconlin $"
__version__ = " $Revision: 684 $"
__date__     = "$Date: 2009-07-07 14:05:36 -0600 (Tue, 07 Jul 2009) $"

"""
This module will mimic the arguments used in the C++ version.  It is provided
for creating simple scripts that will call the C++ programs.
"""

import optparse

class ArgParser(optparse.OptionParser):
    """
    """

    def __init__(self, usage=None):
        if not usage:
            usage = 'usage: %prog [options] args'
        optparse.OptionParser.__init__(self,usage=usage)

        # Add a bunch of options
        self.add_option('-t', '--tracks', help='# Histories to track', 
                dest='tracks', default='50000')
        self.add_option('-i', '--inactive', help='# Inactive Cycles',
                dest='inactive', default='0')
        self.add_option('-a', '--active', help='# Active Cycles', dest='active',
                default='20')
        self.add_option('-b', '--bins', help='# spatial bins for discretization', 
                dest='bins', default='50')
        self.add_option('-w', '--width', help='Width of slab', dest='width', 
                default='20')
        self.add_option('-s', '--seed', help='Random seed', dest='seed',
                default='0')
        self.add_option('-v', '--verbose', help='The verbosity of the output',
                dest='verbose', default='2', type=str)
#       self.add_option('--tol', help='Tolerance', dest='tolerance')
        self.add_option('-f', '--filename', help='Filename for data',
                dest='filename', default='data.dat')
        self.add_option('--run', help='Run simulation', action='store_true', 
                default=False)
        self.add_option('-m', '--multimedia', action='store_false',
                default=True, help='Multi Media problem')


if __name__ == "__main__":
    print "Testing Arguments.py."

    AP = ArgParser()

    op, arg = AP.parse_args()
    print "\noptions entered:"
    for key in op.__dict__.keys():
        print "\t%s: %s" %(key, op.__dict__[key])
    

